[tune] Add AutoMLBoard: Monitoring UI (experimental)

LionHeartWang opened the pull request:
## About AutoMLBoard

 AutoMLBoard can be used as a monitor who collects information about ray tune jobs and
show the trial information to the frontend.

### Install

install django before start AutoMLBoard. run:

```bash
pip install django==1.11.14
```

### Usage

Please refer to `run.py -h` for more help.

```
usage: run.py [-h] --logdir LOGDIR [--port PORT] [--db DB]
              [--reload_interval RELOAD_INTERVAL] [--log_level LOG_LEVEL]

optional arguments:
  -h, --help            show this help message and exit
  --logdir LOGDIR       Directory where AutoML Board will look to find tuning
                        logs it can display
  --port PORT           What port to serve AutoMLBoard on, (default: 8008)
  --db DB               Set SQL database URI in
                        schema://user:password@host:port/database, (default:
                        sqlite3)
  --reload_interval RELOAD_INTERVAL
                        How often the backend should load more data, (default:
                        5)
  --log_level LOG_LEVEL
                        Set the logging level, (default: INFO)
```

As default, sqlite will be used as the database backend engine.
To use mysql as the backend engine, run command like this:

```bash
python run.py --db_address mysql://<hostname>:<port>/<db_name>?user=<user>&password=<password> ...
```

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7240/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7251/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7252/
Test PASSed.

pcmoritz commented:
Thanks for writing this! There is a number of linting related problems you can already fix to speed up the review, see: https://travis-ci.com/ray-project/ray/jobs/138374825

- The Javascript shouldn't be linted clang-format
- There are also some python related linting problems

richardliaw commented:
Thanks for the contribution!

A high level question and comment: will you be providing tests for this codebase?
Also, would it be possible to move this under tune/automlboard?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7299/
Test PASSed.

LionHeartWang commented:
ok,  @richardliaw I've move automlboard to tune/automlboard and will provide tests later on.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7304/
Test PASSed.

richardliaw commented:
Some screen shots would be appreciated here.

<img width="1298" alt="screenshot 2018-08-07 22 09 08" src="https://user-images.githubusercontent.com/4529381/43817506-a48aa982-9a8e-11e8-8429-fb251970653e.png">

richardliaw commented on python/ray/tune/automlboard/backend/collector.py:
```diff
@@ -0,0 +1,371 @@
+            results(list)
+            meta(dict)
+        """
+        for result in results:
```
any reason why these are all class methods?

richardliaw commented on python/ray/tune/automlboard/common/exception.py:
```diff
@@ -0,0 +1,8 @@
+class CollectorError(Exception):
```
we should have a 
```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
```
on all of the files.

LionHeartWang replied:
ok, I've done this~

richardliaw commented on python/ray/tune/automlboard/backend/collector.py:
```diff
@@ -0,0 +1,371 @@
+
+class CollectorService(object):
+    """
+    Server implementation to monitor the status directory and
```
Can you actually run the following to identify all the issues :

```
flake8 ./automlboard --ignore=E121,E123,E126,E226,E24,E704,W503,W504,D100,D401,D202,D104,D105,D107,D209
```

I think you'll need `pip install flake8-docstrings` for this.

LionHeartWang replied:
Thanks for help. I've fixed these issues.

LionHeartWang commented:
hi, @richardliaw  I've commit some improvements and here is some screenshots about the UI.

1. After the server started，there will be a page showing the job status. If the server started in share mode, all jobs's information under logdir will be shown.

![page about jobs](https://raw.githubusercontent.com/LionHeartWang/images/master/tmp/20180808151801.png)

2. you can click the job_id link to see the trials submitted by the job.

![page about trials of a job](https://raw.githubusercontent.com/LionHeartWang/images/master/tmp/20180808151844.png)

3. you can also see the results of each trial reported during the running process.

![page about results of a trial](https://raw.githubusercontent.com/LionHeartWang/images/master/tmp/20180808154416.png)

Killing jobs hasn't been implemented yet in this PR, but we have planned to implement that function in the future. We will continue our work on this and improve the functionality and experience of AutoMLBoard.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7335/
Test PASSed.

robertnishihara commented on python/ray/tune/automlboard/README.md:
```diff
@@ -0,0 +1,6 @@
+AutoMLBoard can be used as a monitor who collect information about ray tune jobs and
+show the trial information to the frontend.
+
+Please refer to `run.py -h` for more help
```
It'd be good to provide more instructions here, e.g., what are the installation dependencies? Does the user need to do `pip install django`?

LionHeartWang replied:
yeah, ``pip install jango`` is needed, I'll add some instructions here.

LionHeartWang commented on python/ray/tune/automlboard/frontend/settings.py:
```diff
@@ -0,0 +1,149 @@
+# https://docs.djangoproject.com/en/1.11/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = 'tktks103=$7a#5axn)52&b87!#w_qm(%*72^@hsq!nur%dtk4b'
```
The secret_key is required to be non-empty. It can be specified by users, I will add some comments here.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7376/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7379/
Test PASSed.

old-bear commented:
Thanks my college @LionHeartWang for this UI board. 
This board is originally built for our AutoML framework which is based on ray.tune. Inside the framework, we use a summary writer to generate more information besides ray.tune, which could be shown by the board. 

However, since the original ray.tune doesn't have such information, these functionalities has to be cutoff in order to contribute the board independently.

So as you can see, now the board only contains a web skeleton with some simple fields. I think this is a good start point. In the future, maybe we could list requirements ordered by priority, and we are willing to implement and contribute continuously.

Features to be implement:
1. Killing job
2. tbd.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7395/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7456/
Test FAILed.

richardliaw commented on python/ray/tune/automlboard/README.md:
```diff
@@ -0,0 +1,16 @@
+## About AutoMLBoard
+
+AutoMLBoard can be used as a monitor who collect information about ray tune jobs and
```
about Tune jobs

LionHeartWang replied:
done~

LionHeartWang commented:
thanks @richardliaw, I will add some commits today to fix the issues you mentioned above.

LionHeartWang commented:
hi, @richardliaw  I've optimized the frontend and here are some screenshots.

1. Summary Page
![page about jobs](https://raw.githubusercontent.com/LionHeartWang/images/master/tmp/20180823150105.png)

2. Page of Trials
![page about trials of a job](https://raw.githubusercontent.com/LionHeartWang/images/master/tmp/20180823150038.png)

3. Page of Results
![page about results of a trial](https://raw.githubusercontent.com/LionHeartWang/images/master/tmp/20180823150133.png)

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7687/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7690/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7692/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7694/
Test PASSed.

LionHeartWang commented on python/ray/tune/automlboard/backend/collector.py:
```diff
@@ -0,0 +1,386 @@
+            time.sleep(self._reload_interval)
+            self._do_collect()
+
+        logging.info("collector stopped.")
```
done~

LionHeartWang commented on python/ray/tune/automlboard/backend/collector.py:
```diff
@@ -0,0 +1,386 @@
+                "trial_id": trial_id,
+                "job_id": job_id,
+                "status": "RUNNING",
+                "type": "RAYTUNE",
```
done

LionHeartWang commented on python/ray/tune/automlboard/frontend/query.py:
```diff
@@ -0,0 +1,111 @@
+
+from django.shortcuts import HttpResponse
+
+from models.models import JobRecord, TrialRecord
```
done~

LionHeartWang commented on python/ray/tune/automlboard/frontend/urls.py:
```diff
@@ -0,0 +1,36 @@
+
+from django.conf.urls import url
+from django.contrib import admin
+from . import view, query
```
done~

LionHeartWang commented on python/ray/tune/automlboard/backend/collector.py:
```diff
@@ -0,0 +1,386 @@
+           meta file.
+
+        Args:
+            job_name(str)
```
done

LionHeartWang commented on python/ray/tune/automlboard/backend/collector.py:
```diff
@@ -0,0 +1,386 @@
+
+JOB_META_FILE = "job_status.json"
+
+EXPR_PARARM_FILE = "params.json"
```
done

LionHeartWang commented on python/ray/tune/automlboard/run.py:
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python
```
done

LionHeartWang commented on python/ray/tune/automlboard/frontend/view.py:
```diff
@@ -0,0 +1,69 @@
+        trials = TrialRecord.objects.filter(job_id=recent_job.job_id)
+        total_num = len(trials)
+        running_num = len(
+            filter(lambda t: t.trial_status == "RUNNING", trials))
```
done~

LionHeartWang commented on python/ray/tune/automlboard/backend/collector.py:
```diff
@@ -0,0 +1,386 @@
+    DEFAULT_LOGDIR = "./ray_results"
+
+    def __init__(self,
+                 log_dir=DEFAULT_LOGDIR,
```
ok, done~

LionHeartWang commented on python/ray/tune/automlboard/frontend/query.py:
```diff
@@ -0,0 +1,111 @@
+    trials = TrialRecord.objects.filter(job_id=job_id)
+
+    total_num = len(trials)
+    running_num = len(filter(lambda t: t.trial_status == "RUNNING", trials))
```
ok, done~

LionHeartWang commented on python/ray/tune/automlboard/backend/collector.py:
```diff
@@ -0,0 +1,386 @@
+    @classmethod
+    def init_logger(cls, log_level):
+        """Initialize logger settings."""
+        logger = logging.getLogger("auomlboard")
```
oh, this statement is not necessary, I've simplified this~

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7701/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7703/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7702/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7707/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7704/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7706/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7708/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7781/
Test PASSed.

LionHeartWang commented:
Design of AutoMLBoard: https://docs.google.com/document/d/1yzyuIw8tIkdW14VveN0Xto3Zu9wmr3dzFBJIc-yw0cg

LionHeartWang commented:
hi @richardliaw, I've made the change required and fix some lint issues. You can review the changes above and let me know if you have any questions. Thank you!

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7819/
Test PASSed.

LionHeartWang commented:
To start a local server for test, install django first.

```bash
pip install django==1.11.14
```

If you don't want to build a new ray wheel package for test, you should copy automlboard directory to your python site-package directory.

```bash
# specify python lib home, /usr/local/lib/python2.7/site-packages, etc
PYTHON_LIB_HOME=/path/to/your/python-site-package

# specify ray log directory, /Users/admin/ray_results, etc
LOGDIR=/path/to/your/logdir

# copy automlboard directory to site-package
cp -r python/ray/tune/automlboard ${PYTHON_LIB_HOME}/ray/tune/

# start service
python ${PYTHON_LIB_HOME}/ray/tune/automlboard/run.py --logdir ${LOGDIR} --port 8008 --reload_interval 10
```

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7824/
Test PASSed.

richardliaw commented:
Hey @LionHeartWang, sorry for the slow reply; I'll review this tomorrow.

richardliaw commented on python/ray/tune/automlboard/templates/index.html:
```diff
@@ -5,119 +5,157 @@
-                             style="width: {{ job.progress }}%;">
-                            <span class="sr-only">{{ job.progress }}</span>
+        <!--<form class="form-inline my-2 my-lg-0">-->
+        <!--<input class="form-control mr-sm-2" type="text" placeholder="Search">-->
```
You'll remove these?

LionHeartWang replied:
done~

richardliaw commented on python/ray/tune/automlboard/templates/job.html:
```diff
@@ -5,103 +5,199 @@
+                            </span>
+                        </div>
+                        <div class="ExperimentView-runs">
+                            <!--form class="ExperimentView-search-controls" style="max-width: 900px">
```
Can you remove these comments?

richardliaw commented on python/ray/tune/automlboard/backend/collector.py:
```diff
@@ -70,15 +60,8 @@ def stop(self):
+                     '%(lineno)d  %(message)s'
         logging.basicConfig(
             stream=sys.stdout, format=log_format, level=log_level)
         return log_level
```
should we return a logger here? and then use the Logger object in the rest of the class?

richardliaw commented:
Got the following error right when I tried it out: 

```
[INFO 2018-08-30 06:08:36,184] collector.py:119  Collector started to run, taking /Users/rliaw/ray_results as parent directory for all job logs.
Performing system checks...

System check identified no issues (0 silenced).
August 30, 2018 - 06:08:36
Django version 1.11.14, using settings 'ray.tune.automlboard.settings'
Starting development server at http://0.0.0.0:8008/
Quit the server with CONTROL-C.
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py", line 328, in execute
    return Database.Cursor.execute(self, query, params)
sqlite3.IntegrityError: NOT NULL constraint failed: models_trialrecord.params

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/Users/rliaw/Research/riselab/ray/python/ray/tune/automlboard/backend/collector.py", line 99, in run
    self._do_collect()
  File "/Users/rliaw/Research/riselab/ray/python/ray/tune/automlboard/backend/collector.py", line 131, in _do_collect
    self.sync_job_info(job_name)
  File "/Users/rliaw/Research/riselab/ray/python/ray/tune/automlboard/backend/collector.py", line 156, in sync_job_info
    self.sync_trial_info(job_path, expr_dir_name)
  File "/Users/rliaw/Research/riselab/ray/python/ray/tune/automlboard/backend/collector.py", line 175, in sync_trial_info
    self._create_trial_info(expr_path)
  File "/Users/rliaw/Research/riselab/ray/python/ray/tune/automlboard/backend/collector.py", line 232, in _create_trial_info
    trial_record.save()
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/models/base.py", line 808, in save
    force_update=force_update, update_fields=update_fields)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/models/base.py", line 838, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/models/base.py", line 924, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/models/base.py", line 963, in _do_insert
    using=using, raw=raw)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/models/query.py", line 1076, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1112, in execute_sql
    cursor.execute(sql, params)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py", line 328, in execute
    return Database.Cursor.execute(self, query, params)
django.db.utils.IntegrityError: NOT NULL constraint failed: models_trialrecord.params
```
Is this expected?

LionHeartWang commented on python/ray/tune/automlboard/static/bootstrap.min.css:
```diff
@@ -1,6 +0,0 @@
-/*!
```
Oh, this is a theme from https://bootswatch.com/united/, a customized version of bootstrap.min.css

LionHeartWang replied:
@richardliaw now I replace the bootstrap.css with bootstrap.min.css and maybe it looks better now.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7880/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7881/
Test PASSed.

LionHeartWang commented:
hi @richardliaw , thank you for the review.

I've addressed the issues left and I've also fixed this error.

```bash
NOT NULL constraint failed: models_trialrecord.params
```

Besides, currently we use python 2.7 with django 1.11.14 and this django version may not fit python3 env. So currently we advise using the board with python 2.7 env, and we will improve the compatibility with python3 in the future.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7891/
Test PASSed.

richardliaw commented:
Errors look unrelated - merging first cut so that we can iterate.

richardliaw commented:
Thanks for your contribution and patience with the process, @LionHeartWang!

LionHeartWang commented:
OK. Thank you @richardliaw  for reviewing!
